### ISSUE TITLE
loading is false, but no data is present (and no error)

### ISSUE THREAD

**Issue by adamdonahue** [apollographql/react-apollo#1314]
Just upgraded to Apollo 2.0, using react-apollo 2.0.0 -- it now appears that the graphql props option for a query is passed a data argument that can have loading set to false but no data.

That is, given a query, say:

```gql
query X($id: ID!) {
  myItem(id: $id) {
    id
    field
  }
}
```

and a `graphql` call:

```javascript
graphql(query, {
  props({ data: { loading, error, myItem } }) {
    console.log('loading =', loading);
    console.log('error =', error);
    console.log('myItem =', myItem);
    return { ... }
 }
})(MyComponent)
```

We're seeing the output:

```
loading = true
error = undefined
myItem = undefined
loading = false
error = undefined
myItem = undefined
loading = false
error = undefined
myItem = {<data>}
```

That is, there is an interim state between `loading` having been set to false but `myItem` yet available on the `data` argument.  

Is this expected, as it appears to differ from the behavior of Apollo 1.0.  And if so, what is the reliable way to check that data was loaded and _is ready for use_ in the component? 

We're now using `(loading || !myItem)` to check whether something is loading.

**Comment by Panoplos** [comment]
I can confirm this issue exists.

![loading false](https://user-images.githubusercontent.com/2549985/32585706-169ffda6-c542-11e7-85cd-4ec15b60298b.png)

![no data](https://user-images.githubusercontent.com/2549985/32585712-1c9f0e2c-c542-11e7-96b7-8efce772dd31.png)

Offending code:

```es6
 graphql(
    gql`
      query localise($locale: LocaleType!, $resourceIds: [String]) {
        l10nSlugs(locale: $locale, resourceIds: $resourceIds) {
          id
          resourceId
          slug
        }
      }
    `,
    {
      name,
      options: ({ locale }) => ({
        variables: {
          locale,
          resourceIds
        },
      }),
      props: props => ({
        l10n: {
          /**
           * Whether or not the graphql resource is loading.
           */
          loading: props[name].loading,
          /**
           * Conveniently offer the locale.
           */
          locale: props.ownProps.locale,
          /**
           * And the BCP47 normalise version.
           */
          bcp47Locale: normaliseBCP47Tag(props.ownProps.locale),
          /**
           * Localise a slug with insertions.
           * @param {string} resourceId Resource ID of the slug to localise.
           * @param {string} slug Default slug in case loading.
           * @param {Array<string> | Object: []} insertions Dynamic content for the slugs.
           * @returns {string} Localised slug.
           */
          localise: (resourceId: string, slug: string, insertions: [string] | Object = []): string => {
            if (!props[name].loading && !props[name].error) {
                           ^^^^^^^^^^ Checked here...
              for (let s of props[name].l10nSlugs) {
                                        ^^^^^^^^^^^ BAM!
                if (s.resourceId === resourceId) {
                  return insert(s.slug, insertions)
                }
              }
            }
            return insert(slug, insertions)
          },
          /**
           * Locale aware ordinals (e.g. 1st, ２番目)
           * @param {Number} ord The numerical ord to convert.
           * @returns {string} The locale-specific ordinal
           */
          ord2str: (ord: Number): string => props.ownProps.l10nShared(`${ord}ord`),
          /**
           * Get whether or not the current locale arranges the name Last, First.
           * @returns {boolean} True if the locale requires a Last, First name arrangement.
           */
          nameLastFirst: (): boolean => NameLastFirstLocales.includes(props.ownProps.locale),

          /**
           * Gets teh minimum number of characters allowed in the locale's name part.
           * @returns {number} The minimum number of characters in the current locale.
           */
          minCharsInNamePart: (): number => MinCharsInNamePart[props.ownProps.locale]
        }
      })
    }
  )
```

**Comment by malexandre** [comment]
Yeah, got the same error when trying to pass my company project to the last version, was previously in 1.4.16.

**Comment by Panoplos** [comment]
Is anyone looking into this? This is a serious bug that is killing quite a bit of my code, now.

**Comment by javascriptlove** [comment]
Can confirm this happens now after i updated today to `apollo-client@2.0.1` and `react-apollo@2.0.1`

**Comment by jbaxleyiii** [comment]
@adamdonahue I'll take a look at this! Still working through issues after launch + took a week off last week! Thanks all for the issue report!

**Comment by javascriptlove** [comment]
Btw, i've resolved that on my side by setting `fetchPolicy` to `cache-and-network`, no more errors after that. So seems like the error is there only with `network-only`

**Comment by Panoplos** [comment]
@javascriptlove Actually, the issue exists with the default setting, too.

**Comment by moimael** [comment]
Any workaround that doesn't require adding a check on data for all loading queries ?

**Comment by peggyrayzis** [comment]
@adamdonahue Can you please provide a stripped down reproduction for us to test? I've tried to reproduce with our test app Githunt, but haven't been able to.

**Comment by adamdonahue** [comment]
Sure, Peggy.  I'll put something together later today.

Adam

**Comment by malexandre** [comment]
I just tested with the 1.0.3 of apollo-client, and everything's work for me. Thanks a lot @jbaxleyiii.

**Comment by jbaxleyiii** [comment]
This has been confirmed as fixed with `1.0.3` in apollo-client! Thank you all for the reporting and helping to diagnose and fix this one!

**Comment by javascriptlove** [comment]
wait wait.. what about the 2.x branch?

**Comment by Panoplos** [comment]
Actually, this is still happening for me with 1.0.3 with the code I shared above. I have confirmed that `l10nSlugs` is `undefined`.

**Comment by Panoplos** [comment]
Something of note, it seems to happen with `graphql` HOCs that depend on the result of other `graphql` HOCs. Like above, `locale` is pushed down from another `graphql` query composition that takes `locale` from Redux, and has a default value, so it always exists.

I have another HOC chain that is something like `account` --> `organisation` -> `services` (where `services` is essentially getting locale-specific labels). It is having this "not loading but data is `undefined`" issue in the `services` query.

**Comment by Panoplos** [comment]
Confirmed fixed in 2.0.1.

**Comment by j4mesjung** [comment]
i am using graphcool service to host my data and all my queries work when I pass in the given ids.  However, on the client side, some return the data and some do not return any data even if `loading: false`.  I am on apollo-client `2.0.4`

**Comment by crokobit** [comment]
still have this issue with version '2.0.4' here 😢 .

**Comment by j4mesjung** [comment]
@crokobit what do you mean by working in safari and failing in chrome?  Are you talking about the `data.loading: false` but not seeing the expected data?

**Comment by Jefffrey** [comment]
I'm on 2.0.4 and the `render` function on a component is called 3 times, 2 of which have loading = `false` and data = `null` and the last one has the data.

<img width="501" alt="screen shot 2017-12-20 at 03 37 30" src="https://user-images.githubusercontent.com/733227/34188534-2a814400-e537-11e7-8776-0482299bab29.png">

What are those extra 2 renderers for?

**Comment by crokobit** [comment]
@j4mesjung yes~

**Comment by ssahni** [comment]
@jbaxleyiii @peggyrayzis This issue still occurs for me. 

I am using react native and

- apollo-cache-inmemory: "^1.1.4"
- apollo-client: "^2.0.4"
- apollo-link-http: "^1.3.2"
- react-apollo: "^2.0.4"

Interestingly enough, the issue only arises when I introduce 'id' as a field in my query. If I do not specify 'id' in my query, then everything works as expected.

Hope this helps!

**Comment by denkristoffer** [comment]
I'm seeing this too. If it helps, I know the data is already in the store.
- "apollo-client": "2.1.0"
- "react-apollo": "2.0.4"

**Comment by xcv58** [comment]
@denkristoffer do you have any reproduce repo? I can't reproduce this in my branch: https://github.com/xcv58/react-apollo-error-template/tree/2591adeeb9401fb90dcb28ed84c44d53ed58285a

**Comment by pencilcheck** [comment]
I have this issue as well, it is a very simple query but it just never return any data when loading is false and error is undefined.

However it returns very quickly when using graphiql

**Comment by jdachtera** [comment]
I have the same problem after running a mutation which returns updated data.

**Comment by felixk42** [comment]
Can confirm I am having the same problem, used to work when id is attached to every queries, but now it doesn't. Seems to be triggered by a mutation that adds something.

**Comment by w9** [comment]
I have the same problem as well.

This is a serious bug and this issue should be re-opened.

**Comment by DennisBaekgaard** [comment]
I got this issue as well, but was able to fix it. In my case it was mostly because the error messages aren't totally clear. I had a mutation which did not return all properties of the entity, this in turn caused the list of entities I was mapping over to be undefined. When I retrieved the last property (albeit being 'null') the data was correctly injected into props. 

This 'gotcha' could probably be either documented or more clearly on the mutation / updating the store pages, or indicated more clearly when there is an error in the console. It seems counter-intuitive when the idea of a graph setup is to retrieve only the properties which are important to you, and not overfetch, when it in turn gives an error in Apollo. At least my 2 cents.

**Comment by finnigantime** [comment]
^ got this same issue, and this workaround worked for me as well (have the mutation return the same complete object subfragment as the original query). Totally unintuitive with graphQL and difficult to manage because different queries in our app return slightly different subfragments. Also no error or warning here, so I just ended up getting `data.foo === undefined` and `data.loading === false` which was a pain to try and debug.

**Comment by davidalekna** [comment]
I've just experienced this issue as well, but only when I test my app on chrome with network throttling, so it may appear when the network is slower or something...?

**Comment by grydstedt** [comment]
This is breaking a lot of things for us. Any known workaround?

**Comment by nadeesha** [comment]
@grydstedt The best workarounds I've found:

1. Include `id`s in the graphql objects whenever you request them. (This helps apollo-client resolve cache better)
2. When you do a mutation , try to return the same graphql selection from mutations, as the query you did.

In no way am I suggesting these are good workarounds. But they are workarounds nonetheless.

**Comment by FezVrasta** [comment]
@nadeesha adding `id` fixed it for us, still annoying tho... Thank for the workaround.

**Comment by Phenominal** [comment]
I am just running into this issue and I don't have the ability to add `id` to every field. Any ideas for an alternate fix? I am using `apollo-boost`. Thanks

**Comment by stevenfitzpatrick** [comment]
Getting this issue as well...the query seems to run fine with loading: false, but the data just remains null even though in network request I can see the query has returned data.

**Comment by pitou** [comment]
The only workaround that "solved" the problem for me was setting `fetchPolicy: 'no-cache'`.
I hope this will help.

**Comment by andfs** [comment]
The error is not in react-apollo, but in apollo-client. I'm using it with angular 4 (apollo-client version 2.2.8) and got the same error. The only workaround was to set fetchPolicy: 'no-cache' just like said @pitou

**Comment by javorosas** [comment]
This issue disappeared for me after migrating from `apollo-client-preset` to `apollo-boost` and `react-apollo@2.1.3`

**Comment by ZiXYu** [comment]
still have the issue, why it was closed? It comes when we try to fetch a query twice in the same time.

**Comment by pierrebiver** [comment]
I do have the same error using `apollo-client 2.2.8` and `apollo-cache-inmemory 1.1.12`.

**Comment by javorosas** [comment]
UPDATE: I'm seeing this strange state again. it occurs after I update the cache using data from an optimisticResponse. I'm thinking that maybe it is caused when you update the same query in the cache simultaneously. Maybe it creates a race condition of some sort.

**Comment by amannn** [comment]
I also just saw this with react-apollo@2.0.4 and apollo-client@2.2.0 (and earlier versions). I'm not able to reproduce it reliably though. Any chance we could re-open this @jbaxleyiii?

**Comment by fuksito** [comment]
I experience this issue with react-apollo 2.1.3 and apollo-boost 0.1.4

**Comment by nicolasg-playster** [comment]
Still having this issue with latest versions, can we re-open this issue?

**Comment by calebdel** [comment]
Experiencing this issue with  react-apollo@2.0.4 and apollo-client@2.2.7

**Comment by steelbrain** [comment]
Fix is up in https://github.com/apollographql/react-apollo/pull/2003

**Comment by adrianschneider94** [comment]
I have the same problem -- I give my component new props, loading is false but the data variable of the render prop function is empty. However, when rendering the first time, it works.
My query is the following:
```
query FeedQuery($first: Int, $offset: Int, $filters: [FilterInput]) {
  datasheets(first: $first, offset: $offset, filters: $filters) {
    id
    productType
    articleId
    excelPath
    weight
  }
}
```

Interestingly, when I exlude the $filters variable, it works. $filters is a list of an input type (FilterInput) which consists of simple scalars. Do array fields need special care?
The query component looks like this:


```
<Query query={FEED_QUERY}
        variables={{first: this.props.first, offset: this.props.offset, filters: this.props.filters}}
        >
    {({loading, error, data}) => {
        if (loading) {return "Loading"}
        if (error) {return "Error"}
        if (data.datasheets.length === 0){return "No result"}
        return (
            <List>
                {
                data.datasheets.map(article =>
                <ListItem key={article.id}>
                    <ArticleCard article={article}/>
                </ListItem>
                )}
             </List>
        )
    }}
</Query>
```

My versions are
* Apollo-boost: 0.1.7
* graphql: 0.13.2
* graphql-tag: 2.9.2
* react-apollo: 2.1.4

**Comment by steelbrain** [comment]
@adrianschneider94 Can you test this with my patch from #2003? You can test by replacing `react-apollo` with `@steelbrain/react-apollo` in your project temporarily

**Comment by adrianschneider94** [comment]
@steelbrain Thanks! The only problem left is that it hangs sometimes in the loading state depite the server returned data. I can‘t really see a pattern but I also don‘t know if the problem is even related.

**Comment by nikogosovd** [comment]
With all the workarounds suggested here, this behaviour is still a bug (48 comments). It is now the main source of pain in process of migration from 1.0 to 2.0.

@jbaxleyiii Are you going to do something with it?

**Comment by felixk42** [comment]
@adrianschneider94 seems related to https://github.com/apollographql/react-apollo/pull/2003#issuecomment-393984128, can't wait for that PR to get merged in...

**Comment by adrianschneider94** [comment]
@felixk42 I already tried @steelbrains workaround. It makes it better, but the query randomly stucks in the loading state.

**Comment by conatus** [comment]
We are also being effected by this bug.

I really hope that either https://github.com/apollographql/react-apollo/pull/2003 or some other fix gets merged soon. Thanks @steelbrain for your work on this.

The situation is as described above in https://github.com/apollographql/react-apollo/issues/1314#issuecomment-352960802 with React on the web.
1. `loading` is `true`, data is not present.
2. `loading` is `false` data is present.
3. Another query is launched.
4. `loading` is `false`, data is absent. React component crashes as a result.

We implemented the two workarounds: 
1. To add `id` to everything requested. This works for the query.
2. To make a mutation return exactly the same fragment that it queries.

As https://github.com/apollographql/react-apollo/issues/1314#issuecomment-370406018 suggests, our expectation was that with GraphQL we should be able to under fetch and things should work as expected. Now we are certainly over fetching with the fragment dragging around a lot of unneeded data into different components. If `id` is so essential to the operation of the cache, it should be added by the client itself without exception.

We'd really appreciate a fix sooner rather than later.

- apollo-client 2.2.8

**Comment by conatus** [comment]
@jbaxleyiii 

Can we please get this opened again? Would be great to see this solved. Happy to test any code to solve it.

**Comment by teoboley** [comment]
Any update on this?

**Comment by kaueburiti** [comment]
Adding the field `id` in *every query from every type* solve the problema to me :)

**Comment by vision669** [comment]
any update on this?

**Comment by ViggoV** [comment]
I also get this with `apollo-boost@0.1.16` and `react-apollo@2.1.11`. The console displays a warning but `console.log(loading, error, result)` gives me `false undefined undefined`. The error in question is `myfield is of type NilClass - MyType expected.`

**Comment by HRK44** [comment]
I switched the ``fetchPolicy`` from ``no-cache`` to ``network-only`` and it works back again for me.

**Comment by pie6k** [comment]
I'd say it's critical bug that introduces a lot of regression in projects. It's here for more than a year I've created tons of code relying on that already and hit the wall now.

**Comment by builtbyproxy** [comment]
My stack is currently: 
- React Native 
- Apollo
- Appsync
- Serverless

A lot of my stack doesn't apply to this issue however, when I debug my react-native application with the use of the React Native Debugger and it's inbuilt inspector: 
![image](https://i.gyazo.com/9b8b58cde85826aa52f77f6fa9cd6943.png)

I have access, inside my appsync object, to `ROOT_MUTATION` and `ROOT_QUERY`. For context, `getRequests` and `openRequest` are the `__typename`'s of the queries/mutations. 

If I clear my cache and then query the HOC, getRequests with, this.props.getRequests it will have no data/getRequests object.

HOWEVER, if I query it with the Query element it will add a `ROOT_QUERY.getRequests` to the appsync object, thus allowing me to forever query it with the previously mentioned HOC. 

This means that if I can find a way to add the query to ROOT_QUERY when I call the HOC, if not already present, I can get data.

Has anyone else run into this use case? 
Hope I helped anyone else, im not to sure how use case it is with Appsync though

**Comment by carlpeaslee** [comment]
We are also encountering this issue and are also exploring various hacky workarounds. Right now there is a theory that this issue is being caused in part by a very large response object.

**Comment by valoricDe** [comment]
In our current project this error only appears if we load 3 components with the same query but different input variables simultaniously.

**Comment by talk2rajeev** [comment]
> still have this issue with version '2.0.4' here 😢 . Sometimes it will work with safari but it constantly failed with chrome.

Same for me.. sometimes it works in safari but constantly fails with chrome..

**Comment by talk2rajeev** [comment]
> In our current project this error only appears if we load 3 components with the same query but different input variables simultaniously.

same for me :(

**Comment by rosskevin** [comment]
Reopened based on #1314.

Issues are more likely to get attention if you PR a breaking test.

**Comment by Torichelly** [comment]
> > In our current project this error only appears if we load 3 components with the same query but different input variables simultaniously.
> 
> same for me 😢

I have the same problem with a few components. They should represent things with a different filtration

Lifecycle is about:
1. loading = true, data = null
2. loading = false, data = myData
3. some magic in other components request
4. loading = false, data = null

**Comment by proteus-aholt** [comment]
This is affecting us, as well.  

Query successful.
Data returned correctly for query when looking at network tab in dev tools.
isLoading=false;
error=undefined;
data={};

**Comment by todesstoss** [comment]
Same here, we are losing data from Apollo cache on some fetches with no error message. Sometimes it helps to add `id` to types on queries if this type of data already fetched by application and `id` was requested on previous fetches. Looks like not documented feature/bug.

**Comment by pie6k** [comment]
My solution for now (which works quite ok) is to not rely at all on "loading" field, but only to check presence of real data. In most of cases it's obviously simple.

Good side-effect is having more smooth data refreshing while loading screen is not flickering around.

**Comment by michelalbers** [comment]
Same here: After a mutation the query data set is empty while loading is false. Added _id and typename to all levels of the mutation and the query.

**Comment by bsotodo** [comment]
The only workaround for me was  to set `fetchPolicy: 'no-cache'`

**Comment by zapo** [comment]
I don't know if someone else can confirm this, but I could see this behavior in my project with apollo-cache-inmemory >= 1.1.12 regardless version of apollo-client & co. As soon as I downgrade to 1.1.11 I no longer have this problem.

**Comment by zcfan** [comment]
After remove the defaultOptions from ApolloClient constructor, errors appear as expected in my project. Hope it helps to locate the problem.

**Comment by valoricDe** [comment]
@zcfan What are the options you are passing to the constructor when the problem does not occur?

**Comment by zcfan** [comment]
@valoricDe Sorry I made a stupid mistake in last comment. After remove defaultOptions error disappeared.

**Comment by Boquete** [comment]
Same problem with apollo-client 2.3.5

Loading works for each fetch policy EXCEPT no-cache.
When no-cache is set, loading is False even if data is still loading.

**Comment by valoricDe** [comment]
Alright guys, we are developers. It can' be that we are not able to fix this. So let's spread some light:
We are doing a mutation somewhere in our app which changes a field defined in a query somewhere differently in our app.
The effect occurs on my machine in the else case of `QueryManager.prototype.getCurrentQueryResult`. There it tries to read the new data from the InMemory cache via `this.dataStore.getCache().read(...)` which results in `{}`.
In the depths of `getCache().read` it goes down until `StoreReader.prototype.diffQueryAgainstStore` which fetches the correct result and saves it to `execResult`. There it says does:
```
if (previousResult) {
            if (isEqual(previousResult, execResult.result)) {
                execResult.result = previousResult;
            }
        }
```
Why do you need to set the result to something which is equal to itself? *strango*

But it does return the correct result. So it must be something on the way back from the store.

**Comment by proteus-aholt** [comment]
in our case, we weren't using caching at all (using internet-only) and were still running into this issue.  I _think_ this is an issue unrelated to caching.

we switched to a different graphql client as it wasn't worth the headache.

**Comment by valoricDe** [comment]
To add to my last post: We are updating the field `id` via a mutation. But are requesting the fields `id` and `name`. So in `QueryManager.prototype.getCurrentQueryResult` we run into the catch case `"Can't find field name on object {
  "id": "58d2498a4b04157110c923",
  "__typename": "Channel"
}."`
Which returns `return { data: undefined, partial: true };`. In `ObservableQuery.prototype.currentResult` we then convert `undefined` to `{}` et voila we have an empty response.

**Comment by valoricDe** [comment]
Looks like it relates to: https://github.com/apollographql/apollo-client/issues/3267 and https://github.com/apollographql/react-apollo/pull/2003

**Comment by Boquete** [comment]
Although, as we occured problem with apollo.loading False, even if data was still loading (no-cache) we found solution (ugly, but still).

On query inside result(data) arrow function, we check data.loading. It seems like this parameter is not related to apollo.loading and works fine . (So we just created variable loading true on mounted page and set false when data is loaded, it might help someone).

But still - it's more a hack then solution.

**Comment by valoricDe** [comment]
As we have already enough hacky solutions in this thread I would advise against posting further hacks and instead contentrate on the problems I reveiled in https://github.com/apollographql/react-apollo/issues/1314#issuecomment-466401693 and https://github.com/apollographql/react-apollo/issues/1314#issuecomment-466410031

**Comment by willtrking** [comment]
This is happening to us as well, in a totally isolated page with no mutations and no other queries. Our structure is retrieving a structure roughly as follows 

`
query Project($uuid: ID!) {
    project(uuid: $uuid) {
    
    milestones(first: 40) {
        edges {
          node {
            
            id
            uuid
            owner
            ...more properties...
            
            parent {
            	id
            	uuid
            	owner
            	...more properties...
          	}
          }
        }
    }
}
`

In this case the `parent` attribute is the same type as the node it's on (a `Milestone`), and that node may or may not be returned from the server already. 

Having ID's on everything does not help. When setting fetchPolicy=`no-cache` we get a different strange behavior, the query is executed 3 times. The first time we have no data (this bug), the second time we have data, woohoo! And the third time we again see this issue

Removing the `parent` attribute immediately solves the problem when fetchPolicy is not set, however the issue persists with the 3 queries as described if fetchPolicy is set

**Comment by valoricDe** [comment]
@willtrking could you have a look if node_modules/apollo-client/core/QueryManager.js: 610 swallows a exception when executing the queries?

**Comment by willtrking** [comment]
@valoricDe 
Looks like a couple things are happening, I'll first address the triple-loading bug when setting `fetchPolicy=no-cache` and then keep looking

With `fetchPolicy=no-cache` the `partial` flag from `QueryManager.prototype.getCurrentQueryResult` is never set when the query isn't done, leading `ObservableQuery.prototype.getCurrentResult` to not set the `loading` flag, and somewhere down the line the `data` field (which is set to undefined in `getCurrentQueryResult`) is being set to an object.


See these lines under `ObservableQuery.prototype.getCurrentResult`

```javascript
var queryLoading = !queryStoreValue ||
            queryStoreValue.networkStatus === NetworkStatus.loading;
        var loading = (this.options.fetchPolicy === 'network-only' && queryLoading) ||
            (partial && this.options.fetchPolicy !== 'cache-only');
```

Changing this to 
```javascript
var queryLoading = !queryStoreValue ||
            queryStoreValue.networkStatus === NetworkStatus.loading;
        var loading = ((this.options.fetchPolicy === 'network-only' || this.options.fetchPolicy === 'no-cache') && queryLoading) ||
            (partial && this.options.fetchPolicy !== 'cache-only');
```

Fixes that issue immediately. I'll come back a bit later with some results from testing WITHOUT `fetchPolicy=no-cache`

**Comment by willtrking** [comment]
@valoricDe 
`QueryManager.prototype.getCurrentQueryResult` is indeed swallowing an exception. In my case it's an `Invariant Violation: Can't find field parent on object` exception. Looks like because SOME of objects returned DO NOT have a `parent` field (as they do not have a parent) while SOME do, Apollo thinks we're missing fields.

Continuing to dig.

**Comment by adrianos10** [comment]
I have that issue too on the latest version of `apollo-client`.  Removing `fetchPolicy: 'no-cache'` from query options fixes the problem. I have reverted `apollo-client` version to `2.4.13` and it seems to be working properly.

**Comment by mattbiltoneagle** [comment]
Also having this issue

I'm not sure there's an issue within `apollo-client` because I use a loggerLink and it reports that the query did resolve with the expected data.

Could this possibly be related to the `componentWillReceiveProps` method in the `Query` component?

Not sure what the cause was but this method was behaving very inconsistently for me in my own components. Since it's being deprecated I updated them to use `getDerivedStateFromProps` and `componentDidUpdate` (as is recommended in the official react guide) and that fixed things.

**Comment by mattbiltoneagle** [comment]
UPDATE

I've narrowed down the location of the bug. It appears to be in the cache.

In `DataStore.markQueryResult` ([this file](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/data/store.ts)) I receive the data I expect, and then it places the data in the cache:

```javascript
this.cache.write({
        result: result.data,
        dataId: 'ROOT_QUERY',
        query: document,
        variables: variables,
      });
```

However, when the result is assembled further down the line (`QueryManager.getCurrentQueryResult` [here](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/QueryManager.ts)) it's drawn from the cache again, but the cache doesn't find anything:

```javascript
try {
        // the query is brand new, so we read from the store to see if anything is there
        const data =
          this.dataStore.getCache().read<T>({
            query,
            variables,
            previousResult: lastResult ? lastResult.data : undefined,
            optimistic,
          }) || undefined;

        return { data, partial: false };
      } catch (e) {
        // I end up here. Signalling that there's nothing in the cache
        return { data: undefined, partial: true };
      }
```
I'm assuming then that the issue is possibly related to the way the cache is generating keys for results?

Result definitely went in... but result did not come back out...

**Comment by jonathanhuet** [comment]
I also have the same issue with the `fetchPolicy: 'no-cache'` then the value cycle of the `loading` property of `Query` was `false` -> `true` -> `false` in version 2.5.1. I went back to version 2.4.13 as suggested by @adrianos10  (thanks ;) ) and then all worked correctly

**Comment by mattbiltoneagle** [comment]
Possible solution for everyone here...

Are you updating all of your state immutably? I just updated all of my reducers/methods-making-calls-to-setState and made sure that any operations which touch broken components do a deep clone of the state.

This fixed it for me

This tells me that cache diffs are based on object ids rather than the primitives within them.

Immutable state updates are obviously a good thing.... but they're also expensive and are an 'opt-in' thing in javascript (either with libraries like `immutable` or by being a smart-alec).

What should change to stop more people falling into this hole? Should the docs warn against non-immutable state updates... or does the cache diffing need to change?

**Comment by imakou** [comment]
The issue is still there in ver 2.5.3.

The pic shows that the data has been fetched but the component has been re-rendered again.
so, that it causes an empty returning...
 
![image](https://user-images.githubusercontent.com/1507950/54487566-ee68ee00-486d-11e9-9ba1-7352d6fc816e.png)

Any suggestions?

**Comment by boboci9** [comment]
Hi, 
I have the same issue loading comes back false but data and error are both undefined when I first press the button, if I press the button again the error field will be populated because we don't allow the same id, and if I change the id before pressing it for the first time then I get loading=false and data=populated correctly but never on the first push of the button. I am pretty new with react so it is possible I am doing something wrong but here is my code. Any input is welcome.
"react-apollo": "^2.5.3", // I tried it with 2.0.8 too upon previous suggestion but same issue

```

export const ADD_ENTRY_TO_LIST_ORG = gql`
  mutation addEntryToOrganizationList(
    $_id: ObjectID
    $field: String
    $entry: ListEntryInput
  ) {
    addEntryToOrganizationList(_id: $_id, field: $field, entry: $entry) {
      _id
      name
      termsAndConditions
      program {
        _id
        innerId
        name
      }
    }
  }
`;
 <Mutation mutation={ADD_ENTRY_TO_LIST_ORG}>
              {(addTerms, { loading, error, data }) => (
                <div>
                  <form
                    onSubmit={e => {
                      e.preventDefault();
                      addTerms({
                        variables: {
                          _id: this.state.organization._id,
                          field: this.state.field,
                          entry: {
                            innerId: idInput.value,
                            name: nameInput.value
                          }
                        }
                      });
                      console.log("updated", loading, error, data);
                      if (data && data.addEntryToOrganizationList) {
                        this.state.current =
                          data.addEntryToOrganizationList[this.state.field];
                      }
                      if (error) {
                        console.log("error", error.message);
                        $.notify(error.message, { type: "danger" });
                      }
                    }}
                  >
                    <div className="form-group">
                      <label>Id</label>
                      <input
                        id="id"
                        type="text"
                        className="form-control"
                        ref={node => {
                          idInput = node;
                        }}
                      />
                    </div>
                    <div className="form-group">
                      <label>Name</label>
                      <input
                        id="name"
                        type="text"
                        className="form-control"
                        ref={node => {
                          nameInput = node;
                        }}
                      />
                    </div>

                    <button type="submit" className="btn-green">
                      Add new
                    </button>
                  </form>
                </div>
              )}
            </Mutation>
```

**Comment by emanuelschmitt** [comment]
Encountering a similar error in 2.5.1. 
The data coming from the `Query`-component is `{}` instead of undefined and then switches to the actual expected payload.

<img width="479" alt="Screenshot 2019-03-27 at 17 09 37" src="https://user-images.githubusercontent.com/4573648/55092801-84451a00-50b3-11e9-9a0a-f9d927949e7e.png">

**Comment by smkhalsa** [comment]
@imakou are you querying for the `ID` for "list" in you query above? I had the same issue, and what solved it for me was ensuring that I was querying the `ID` field for all subqueries.

**Comment by Nabuk3** [comment]
@smkhalsa Similar problem on my side - resolved by adding ID in all query entities even those never touched in cache operations. Thanks.

**Comment by cosmarc** [comment]
Same issue with:
"react-apollo": "^2.5.5"
"apollo-client": "^2.5.1"

**Comment by cosmarc** [comment]
I've just found out that I don't have the issue anymore if I disable cache persistence (I am using apollo-cache-persist)

**Comment by andfs** [comment]
I solved putting: 

**errorPolicy: "all"**

**Comment by dakhipp** [comment]
I'm sure it won't help all cases of where this is happening, but this was happening for me when resolving an object with undefined values instead of null values.

Example:
```
// works
res.json({
  contact: contact || null,
  hasPaid: false
});

// doesn't work
res.json({
  contact, // contact: undefined 
  hasPaid: false
});
```

**Comment by scientist1642** [comment]
This was really confusing, spent hours. Suggestions above to query ids of all subfields solved it but is it required by Apollo to work?

Somehow a query in one location caused the other one to re-execute and later ended up with `{loading:false, data:{}, error:undefined}`

**Comment by hwillson** [comment]
`react-apollo` 2.5.6 includes new `returnPartialData` functionality, that will help address this issue. Please see the [changelog](https://github.com/apollographql/react-apollo/blob/master/Changelog.md#256-2019-05-22) for more details. Thanks!

**Comment by valoricDe** [comment]
@hwillson I vote for reopening this issue. As, even without the returnPartialData flag, the apollo Query component should not return an empty `{}` when some required data is not found.

**Comment by hwillson** [comment]
@valoricDe Setting `returnPartialData` to `true` is the solution for this until the next major release of Apollo Client. After that, it will become the default.

**Comment by amannn** [comment]
@hwillson Thanks for the update on this issue!

> Setting returnPartialData to true is the solution for this until the next major release of Apollo Client. After that, it will become the default.

Does that mean that in the next major release, in the [example from the changelog](https://github.com/apollographql/react-apollo/blob/master/Changelog.md#256-2019-05-22), the component which uses `GET_MEMBER_WITH_PLANS` would resolve immediately with some properties missing _by default_?

If so, that sounds somewhat dangerous to me. If the second component is implemented to expect all properties to be available, it would work if it is implemented first. As soon as you add a component that requests less fields, the previous component could break.

Also: Is returning partial data really the fix to the problem described at the top of the issue? To me, a good default would be that `loading` becomes `false` as soon as all necessary data for a given query is fetched and that data can be provided to the component – no matter if another component uses a similar query. Handling partial data should be opt-in IMO, as it requires additional null checks, even for properties which are non-nullable in the schema.

Or am I missing something?

**Comment by proteus-aholt** [comment]
I don't see how `returnPartialData` has anything to do with the original bug.  The issue here wasn't that we had extra web requests or anything that `returnPartialData` addresses.  The issue here was (stolen from one of the original comments on this issue):
```
Render flow goes like this:

loading: true, no data
loading: false, data is available and good
loading: false, no data
Hope this helps!
```

from your changelog:
```
Apollo Client will not re-use the partial data that was cached from the first query, when it preps and displays the second component. It can't find a cache hit for the full second query, so it fires the full query over the network.
```

Sorry, but not gonna help for this bug.  If anything this feature will just make this bug harder to reproduce.  It is a band-aid over the real problem.

**Comment by hwillson** [comment]
Would someone here be able to provide a small runnable reproduction that clearly demonstrates this issue? Talking around a concrete example of the problem will definitely help.

**Comment by amannn** [comment]
I was able to reproduce the issue, but the use case is a bit special. In my case it was related to how cache identifiers are generated. The example is a bit contrived, but maybe it helps to debug the issue.

Have a look at this code sandbox: https://codesandbox.io/s/apollo-error-reproduction-scrcx.

One of the query components always renders an empty object. If you remove one, the other one always succeeds.

If you look at the cache, it becomes clear why this fails:

```json
{
  "Article:5d010e4e0bcb900018dd227b": {
    "id": "5d010e4e0bcb900018dd227b",
    "author": {
      "type": "id",
      "generated": false,
      "id": "Author:Holden Luntz Gallery",
      "typename": "Author"
    },
    "__typename": "Article",
    "href": "/article/holden-luntz-gallery-chapter-30-dialogues-great-photographers"
  },
  "Author:57a4ab99b202a366110001a9": {
    "id": "57a4ab99b202a366110001a9",
    "profile_handle": null,
    "__typename": "Author"
  },
  "ROOT_QUERY": {
    "articles({\"sort\":\"PUBLISHED_AT_ASC\"})": [
      {
        "type": "id",
        "generated": false,
        "id": "Article:5d010e4e0bcb900018dd227b",
        "typename": "Article"
      },
      // ...
    ],
    "article({\"id\":\"5d010e4e0bcb900018dd227b\"})": {
      "type": "id",
      "generated": false,
      "id": "Article:5d010e4e0bcb900018dd227b",
      "typename": "Article"
    }
  },
  "Author:Holden Luntz Gallery": {
    "id": "57a4ab99b202a366110001a9",
    "name": "Holden Luntz Gallery",
    "__typename": "Author"
  }
}
```

The entry with the identifier `Author:57a4ab99b202a366110001a9` is no longer referenced in the cache and therefore the component that needs the data from this entry fails. I'm not sure though what the ideal solution would be as this is due to the normalization.

If you change the `returnPartialData` at the top, both components render, but one is missing data.

I had this bug a few times in a project where cache identifiers might be different based on what data is fetched / available.

I have to say this is a special use case. Maybe someone else has a better example? Maybe the comments above provide a good example.

**Comment by kgstew** [comment]
I found a simple example that may help illustrate the issue. 

`data` will come back as an empty object if trying to fetch a field using the `@client` directive for a field that does not yet exist in the cache or have its own client resolver. There are no errors just an empty `data` object.

Removing the directive and trying to fetch a field that does not exist on the server does not cause the same issue however and the query will return the fields that do exist, ignoring the missing field.

The example is here.
https://codesandbox.io/embed/github/kgstew/react-apollo-bug-testing/tree/testing_client_directive/

**Comment by EternalVigil** [comment]
I've been experiencing a very similar issue when creating a React application with hooks.  I have one component nested inside the other and each component has an independent query associated with it; it seems to resolve the outside component but returns an empty object for the inner component.  I've test each component as a stand-alone and they both resolve fine when it's just the single query.  Any ideas?

**Comment by amannn** [comment]
@EternalVigil can you create a reduced CodeSandebox example?

@hwillson can we please re-open this issue? It doesn't seem to me that this is solved.

**Comment by codenakama** [comment]
I am experiencing the same issue as well. I have a Dropdown component in the middle of the render of my component, some data has to be fetched for the dropdown options. after selecting an option, the query reads from the cache but the value of the field inside data is undefined.

**Comment by Oliver-ke** [comment]
I'm having the same issue, everything worked fine not until I added routing with react-router-dom and boom everything crashes, queries are only made on the first render,  and once props changes query now returns data as empty object {}

**Comment by dylanwulf** [comment]
@shyamzenefits have you tried upgrading react-apollo? There was an issue with `error` disappearing which was fixed in v3.1.0

**Comment by shyamzenefits** [comment]
@dylanwulf Thanks for the swift response . Will try upgrading 👍

**Comment by shyamzenefits** [comment]
@dylanwulf Tried upgrading the apollo packages to v3.1.0 , but still facing the same issue . Any Suggestions ?

**Comment by dylanwulf** [comment]
@shyamzenefits Try upgrading to the latest available version (I think it's 3.1.3 right now). If that doesn't help, create a runnable reproduction (possibly with codesandbox) and post it in a new issue.

**Comment by japrogramer** [comment]
I have the same issue, in react data is always undefined despite receiving data from the server

    "react-apollo": "^3.1.3",
    "apollo-boost": "^0.4.4",
    "apollo-link-ws": "^1.0.14",
    "apollo-upload-client": "^11.0.0",


```
      <Query
        query={TimelineQuery}
        variables={{ page: page}}
        fetchPolicy={'network-only'} // skip the cache
      >
        {({ loading, error, data, refetch }) => {
          if (loading || !data) return <Linear />;
          console.log(loading, error, data, refetch)
          if (data == null) return <Linear />;
          return (
            <React.Fragment>
              <Sauce data={data}/>
            </React.Fragment>
          );
        }}
      </Query>

```

```
query TimeLine($page: Int) {
  currentTimeline(page: $page, perpage: 10) {
    pageInfo {
      Count
      CurrentPage
      NumPages
      hasNext
      hasPrevious
      hasOtherPages
      nextPageNumber
      previousPageNumber
      startIndex
      endIndex
      __typename
    }
    edges {
      node {
        subject {
          ...AnnouncementsInfo
          ... on PanelType {
            id
            hidden
            createdAt
            zipCode
            likesCount
            liked
            followsCount
            followed
            group {
              name
              id
              __typename
            }
            categories {
              edges {
                node {
                  id
                  category
                  __typename
                }
                __typename
              }
              __typename
            }
            comments {
              id
              createdAt
              liked
              likesCount
              thanked
              thanksCount
              attachments {
                attachment
                previewSm
                previewMd
                __typename
              }
              
              __typename
            }
            __typename
          }
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }
}

fragment AnnouncementsInfo on AnnouncementType {
  
  categories {
    edges {
      node {
        id
        category
        __typename
      }
      __typename
    }
    __typename
  }
  attachments {
    attachment
    previewSm
    previewMd
    __typename
  }
  group {
    name
    id
    __typename
  }
  genComments {
    ...genComments
    __typename
  }
  createdAt
  modifiedAt
  hidden
  zipCode
  title
  id
  likesCount
  liked
  __typename
}

fragment genComments on GenCommentsType {
  id
  createdAt
  liked
  likesCount
  thanked
  thanksCount
  attachments {
    attachment
    previewSm
    previewMd
    __typename
  }
  author {
    id
    username
    fullName
    institution
    position
    profilePicSm
    __typename
  }
  __typename
}
```

```
{
  "data": {
    "currentTimeline": {
      "pageInfo": {
        "Count": 261,
        "CurrentPage": 1,
        "NumPages": 27,
        "hasNext": true,
        "hasPrevious": false,
        "hasOtherPages": true,
        "nextPageNumber": 2,
        "previousPageNumber": null,
        "startIndex": 1,
        "endIndex": 10,
        "__typename": "PaginatorInfo"
      },
      "edges": [
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1NQ==",
              "hidden": false,
              "createdAt": "2019-11-14 11:19:52",
              "zipCode": "77098",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjQ5NQ==",
                      "category": "Mental health",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1NA==",
              "hidden": false,
              "createdAt": "2019-11-14 10:58:58",
              "zipCode": "Salado",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjk0Mw==",
                      "category": "sa",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1Mw==",
              "hidden": false,
              "createdAt": "2019-11-06 14:34:44",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [
                {
                  "id": "Q29tbWVudHNUeXBlOjc1MQ==",
                  "createdAt": "2019-11-13 14:11:40",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                }
              ],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1Mg==",
              "hidden": false,
              "createdAt": "2019-11-04 22:17:32",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1MQ==",
              "hidden": false,
              "createdAt": "2019-11-01 14:19:29",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI1MA==",
              "hidden": false,
              "createdAt": "2019-11-01 00:02:25",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI0OQ==",
              "hidden": false,
              "createdAt": "2019-10-31 13:05:12",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI0OA==",
              "hidden": false,
              "createdAt": "2019-10-30 17:48:15",
              "zipCode": "Baytown",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxNw==",
                      "category": "mental health, depression, ptsd",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjIzNg==",
              "hidden": false,
              "createdAt": "2019-09-19 07:49:24",
              "zipCode": "77098",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": {
                "name": "Houston Mental Health Therapy Coalition",
                "id": "R3JvdXBUeXBlOjEx",
                "__typename": "GroupType"
              },
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjI5NQ==",
                      "category": "Counseling",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  },
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjQ2Mw==",
                      "category": "Domestic violence",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [
                {
                  "id": "Q29tbWVudHNUeXBlOjc0MQ==",
                  "createdAt": "2019-10-25 23:33:40",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjc0Mg==",
                  "createdAt": "2019-10-25 23:33:45",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjc0Mw==",
                  "createdAt": "2019-10-25 23:33:50",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjc0NA==",
                  "createdAt": "2019-10-25 23:33:57",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                }
              ],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        },
        {
          "node": {
            "subject": {
              "id": "UGFuZWxUeXBlOjI0NQ==",
              "hidden": false,
              "createdAt": "2019-10-14 11:15:41",
              "zipCode": "Houston",
              "likesCount": 0,
              "liked": false,
              "followsCount": 0,
              "followed": false,
              "group": null,
              "categories": {
                "edges": [
                  {
                    "node": {
                      "id": "Q2F0ZWdvcnlUeXBlOjkxOA==",
                      "category": "mental",
                      "__typename": "CategoryType"
                    },
                    "__typename": "CategoryTypeEdge"
                  }
                ],
                "__typename": "CategoryTypeConnection"
              },
              "comments": [
                {
                  "id": "Q29tbWVudHNUeXBlOjcyMQ==",
                  "createdAt": "2019-10-19 01:51:42",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjcyMg==",
                  "createdAt": "2019-10-22 13:29:10",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjczOQ==",
                  "createdAt": "2019-10-25 20:02:00",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                },
                {
                  "id": "Q29tbWVudHNUeXBlOjc0MA==",
                  "createdAt": "2019-10-25 20:02:05",
                  "liked": false,
                  "likesCount": 0,
                  "thanked": false,
                  "thanksCount": 0,
                  "attachments": [],
                  "__typename": "CommentsType"
                }
              ],
              "__typename": "PanelType"
            },
            "__typename": "ActivityType"
          },
          "__typename": "TimelinePQEdge"
        }
      ],
      "__typename": "TimelinePQConnection"
    }
  }
}
```

**Comment by vjsingh** [comment]
I just tried querying a list of fields with an 'id' param by mistake. Instead of getting an error I just got no data.

**Comment by jcace** [comment]
Same issue here, using apollo-boost 0.4.7 and @apollo/react-hooks 3.1.3. Basically, my app will end up with `loading: false, error: undefined, data: undefined`. Has this really not been fixed yet?

**Comment by amannn** [comment]
Can we reopen this issue? I saw this issue in two apps recently. I couldn't reproduce the bug so far but it's definitely still happening with the latest versions.

**Comment by schonfeld** [comment]
FWIW, I've experienced this issue and tracked it down to my query asking for a client schema local field that was no longer available. Removing the reference to that local field fixed the issue.

For example, I'd have a query such as this:
```graphql
{
query Thread() {
  thread {
    messages {
      context {
        ... on Event {
          id
          state
          isPast @client
        }
      }
    }
  }
}
```

That `isPast @client` was recently removed from our client schema. Once I removed that field from the request, my `useQuery` calls started coming back with data, without needing to make use of `returnPartialData`.

Hope this helps!

**Comment by gamsterx** [comment]
I just stumbled upon the same issue. I can confirm that it only appears with  `network-only`. I'm using the angular client so maybe this is not only react-apollo related?

In my query the roles array objects are `undefined` event though the response from the server contains the individual role objects...
 
```graphql
 query getAreaById($id: uuid!) {
      area: areas_by_pk(id: $id) {
        id
        name
        roles {
          role {     // <== this object in the roles array is undefined
            id
            type
          }
        }
      }
    }
```

**Comment by KyleLee95** [comment]
Seems like this issue is being revived so I thought I would chime in as well.

Context:
Using Apollo-Client + react-apollo to connect to Shopify's Storefront API

Versions:
react-apollo: 3.1.5
apollo-client: 2.6.8

What I'm querying:



    shop {
      name
      description
      products(first: 20) {
        pageInfo {
          hasNextPage
          hasPreviousPage
        }
        edges {
          node {
            id
            title
            options {
              id
              name
              values
            }
            variants(first: 250) {
              pageInfo {
                hasNextPage
                hasPreviousPage
              }
              edges {
                node {
                  id
                  title
                  selectedOptions {
                    name
                    value
                  }
                  image {
                    src
                  }
                  price
                }
              }
            }
            images(first: 250) {
              pageInfo {
                hasNextPage
                hasPreviousPage
              }
              edges {
                node {
                  src
                }
              }
            }
          }
        }
      }
    }
  
So far I've been able to get my data to render by doing this:
```
   <Query query={test}>
          {(res, loading, error) => {
            if (loading || !res.data) return 'loading...'
            if (error) return `${error}`
            return <React.Fragment>{res.data.shop.name}</React.Fragment>
          }}
    </Query>
```
Handling loading and !res.data the same way (since data is undefined by default). However, this is probably not the best way to handle this since res.data could truly be returning undefined.

**Comment by leonvdb** [comment]
I seem to have the a similar issue, but I'm not sure if it's exactly the same.

- Doesn't happen for me consistently. In 80% of the cases the query works fine and I have data. It happens randomly. Data and environment always exactly the same.. 
- When the problem does occur, I actually can see in my network that I receive the data, but the data object I get from react-apollo is undefined
- fetch-policy doesn't make a difference for me

Is this similar for anyone?

**Comment by joselrodrigues** [comment]
Hi, i had same issue.
Context:
Using Apollo-Client + react-apollo to connect to django-graphene API.

I try to get jwt token from my back with:

const LOGIN = gql`
      mutation TokenAuth($username: String!, $password: String!) {
        tokenAuth(username: $username, password: $password) {
            token
            payload
            refreshExpiresIn
         }
      }
    `;
const [updateLogin, { loading, error }] = useMutation(LOGIN);

What i did to solve my problem was:

updateLogin({ variables: { username:"admin",  password:"admin"}}).then((data) => {
       console.log(data)
 })

it work for me every time.

**Comment by kierangillen** [comment]
Confirmed that adding `id` for every applicable field will fix this issue.

**Comment by art1373** [comment]
my query
```

**Comment by art1373** [comment]
my query ```{
    users {
      data {
        name
        email
      }
    }
  }
```
no ID or anything. on react native version 61.5 on **ios** there is no problem, but on Android, although the URI is https nothing is happening no loading becomes true and no error or data.

**Comment by art1373** [comment]
> ```
> my query{
>    users {
>      data {
>        name
>        email
>      }
>    }
>  }
> ```
> 
> no ID or anything. on react native version 61.5 on **ios** there is no problem, but on Android, although the URI is https nothing is happening loading becomes true and no error or data.

The problem was with t android emulator I switched to a real device and the query came back fine with data
